HIP/hip-1195.md

```diff
@@ -0,0 +1,834 @@
+hook's EVM bytecode and a default gas limit for its execution. The `LambdaEvmHook` message also includes the initial
+storage slots for a lambda hook, if desired.
+
+```protobuf
+/**
+ * Definition of a lambda EVM hook.
+ */
+message PureEvmHook {
+  /**
+   * The specification for the hook.
+   */
+  EvmHookSpec spec = 1;
+}
+
+/**
+ * Definition of a lambda EVM hook.
+ */
+message LambdaEvmHook {
+  /**
+   * The specification for the hook.
+   */
+  EvmHookSpec spec = 1;
+
+  /**
+   * Initial storage contents for the lambda, if any.
+   */
+  repeated LambdaStorageSlot storage_slots = 2;
+}
+
+/**
+ * Shared specifications for an EVM hook. May be used for any extension point.
+ */
+message EvmHookSpec {
+  /**
+   * The source of the EVM bytecode for the hook.
+   */
+  oneof bytecode_source {
+    /**
+     * The id of a contract that implements the extension point API with EVM bytecode.
+     */
+    ContractID contract_id = 1;
+  }
+
+  /**
+   * If present, the default gas limit to use when executing the EVM hook.
+   */
+  google.protobuf.UInt64Value default_gas_limit = 2;
+}
+
+/**
+ * A slot in the storage of a lambda EVM hook.
+ */
+message LambdaStorageSlot {
+  /**
+   * The 32-byte key of the slot; leading zeros may be omitted.
+   */
+  bytes key = 1;
+
+  /**
+   * If the slot is present and non-zero, the 32-byte value of the slot;
+   * leaving this field empty or setting it to binary zeros in an update
+   * removes the slot.
+   */
+  bytes value = 2;
+}
+```
+
+The indexes of newly installed hooks will appear in the legacy `TransactionReceipt` if records streams are enabled,
+
+```protobuf
+
+message TransactionReceipt {
+  // ...
+
+  /**
+   * In the receipt of a successful create or update transaction for an entity that supports hooks,
+   * the indexes of any newly installed hooks.
+   */
+  repeated uint64 installed_hook_indexes = 16;
+}
+```
+
+Once a hook is installed to an entity, a transaction generally references it by index relative to an implicit owner.
+The details of the call are specified based on its type; for example, EVM hook calls are specified by an `EvmHookCall`
+message that gives optional call data and gas limit.
+
+If the called hook does not match the given call specification, the network will fail the transaction with
+`BAD_HOOK_REQUEST`. If there is no hook installed at the specified index, the network will fail the transaction
+with `HOOK_NOT_FOUND`.
+
+```protobuf
+/**
+ * Specifies a call to a hook from within a transaction where
+ * the hook owner is implied by the point of use. (For example,
+ * it would never make sense to try to use an account allowance
+ * hook for account 0.0.X inside an AccountAmount for account
+ * 0.0.Y; hence we only need to give the index of which of
+ * 0.0.Y's hooks we want to call.)
+ */
+message HookCall {
+  /**
+   * The index of the hook to call.
+   */
+  uint64 index = 1;
```

Comment:
Instead of assuming and computing the value of the `AccountID` on a case-by-case basis, I suggest using the `HookID` that contains it. This would make things simpler for developers by delegating the definition of the `AccountID` to the SDK.

Reply:
Notice that the `AccountID` of the hook installer must already be in the [`AccountAmount`](https://github.com/hiero-ledger/hiero-consensus-node/blob/main/hapi/hedera-protobuf-java-api/src/main/proto/services/basic_types.proto#L444) or [`NftTransfer`](https://github.com/hiero-ledger/hiero-consensus-node/blob/main/hapi/hedera-protobuf-java-api/src/main/proto/services/basic_types.proto#L501) message.

The same will hold for all hooks; they will always be executed in a context that explicitly references the installing entity.